cores/arduino/wiring_digital.c

```diff
@@ -26,6 +26,8 @@
 #include "wiring_private.h"
 #include "pins_arduino.h"
 
+const uint8_t digital_pin_count = sizeof(digital_pin_to_port_PGM) / sizeof(digital_pin_to_port_PGM[0]);
```

Comment:
It's only one line of code, but I think there are a couple of problems with it :-) :

1) This creates an external linkage for `digital_pin_count`. The compiler will actually allocate one byte to this variable (Otherwise, your `extern const uint8_t digital_pin_count` in the `pins_arduino.h` would fail). However, this is a compile-time constant. We don't need to allocate storage for it.  I think this needs to be moved into a `.h` file, and written with the `static` keyword (I'm not sure which one, but probably where it is used?)

```C++
static const uint8_t DIGITAL_PIN_COUNT = 
    sizeof(digital_pin_to_port_PGM) / sizeof(digital_pin_to_port_PGM[0]);
```
so that it has internal linkage, and the compiler can optimize it out.

2) Since it's a compile-time constant, I think it should be capitalized? ie. `DIGITAL_PIN_COUNT`.

Reply:
1. LTO gets rid of that variable. Looking through the generated assembly, not only can I not find a variable with that name, I cannot find a piece of memory that is given the value 0x14 (20 - the value of that variable for an ardiuno).
2. No complaints per se. I Are there coding standards for this project? I always through all caps was for macros. Why are the arrays like `port_to_mode_PGM` not all caps?
3. Have you checked? As in point (1), I cannot find the variable, and the generated code doesn't look like it's using it.

Reply:
But you are right, I would rather not rely on the optimiser for that, and I would rather have it as a `constexpr` in a header file. The issue is that `digital_pin_to_port_PGM` is defined in `pins_arduino.h`, while `digitalPinToPort` is defined in `Arduino.h`. I would rather not force Arduino.h to include pins_arduino.h, particularly for a minor or non-existent optimisation like that.

Reply:
Heh, I guess LTO saves us from a lot of sloppy programming. :-) Eventually, this code may get copied to other Arduino Cores using other architectures with their own compilers (ARM, ESP8266, ESP32, etc). Can you be sure that LTO will work correctly for those compilers? Personally, I always feel better having internal linkage for compile-time constants. Then you don't have to depend on LTO for the compiler to get rid of the storage.

Reply:
I don't know if I'd call it sloppy. It's correct in terms of language, intent and execution. It's easy to read and not overly complex. It doesn't change the existing code base more than it needs to.

To save this hypothetical one byte for possible future implementations that don't have an optimising compiler, where do you think this definition should go?

Reply:
I regret the use of "sloppy", it was not diplomatic.

Usually, I think the definition of the size of a variable-sized array (the `digital_pin_to_port_PGM`) should go right after its definition. In other words, in the various `pin_arduino.h` files.

It looks like you will have to include an
```C++
extern const uint8_t digital_pin_count;
```
in each of those files anyway. It seems better to insert:
```C++
static const uint8_t DIGITAL_PIN_COUNT =
    sizeof(digital_pin_to_port_PGM) / sizeof(digital_pin_to_port_PGM[0]);
```
instead. I realize it's slightly more code duplication, but it seems to be more direct and clear to me.

Reply:
I looked more into it, and it can't be done how you want it.

For all translation units except `wiring.c` the arrays are `extern`. The definition of the arrays in `pins_arduino.h` is guarded by `#ifdef ARDUINO_MAIN`, which is only defined in `wiring.c`. This makes sense: If you're concerned about the single byte of their size and you don't believe that your compiler can optimise, you should be far more concerned to make sure that the arrays themselves aren't duplicated.

If the array is `exern`, then you can't `sizeof(...)` it. So the variable that depends on `sizeof(...)` must be `extern` too.

The `extern const uint8_t digital_pin_count` variable should be declared just after the extern arrays, in `Arduino.h`, which will make it more obvious why it must be `extern` and also resolve your comment regarding the definition in only one `pins_arduino.h`.

Even if it could be done, I don't think it's worthwhile and don't want to spend my time trying to optimise for a single byte that the linker takes care of anyway. Can I request that you rather not oppose this, then make a pull request along the lines of 'optimise bounds checking of digital pin functions` where you can make the optimisations as you see fit? Or you can request that I close this PR and you can make one that does everything optimised as you please; I would comply with that.